optaplanner-core/src/main/java/org/optaplanner/core/impl/solver/DefaultSolverJob.java

```diff
@@ -223,7 +230,10 @@ public Duration getSolvingDuration() {
     private final class UnlockLockPhaseLifecycleListener extends PhaseLifecycleListenerAdapter<Solution_> {
         @Override
         public void solvingStarted(SolverScope<Solution_> solverScope) {
-            solverStatusModifyingLock.unlock();
+            // The solvingStarted event can be emitted as a result of addProblemChange().
+            if (solverStatusModifyingLock.isLocked()) {
+                solverStatusModifyingLock.unlock();
+            }
```

Comment:
I am not up to date with the locking code any more (even before this change). cc @triceo
Who groks it (how confident are we that it's fine) and how good is our test coverage?

Reply:
That's irrelevant of this change and this PR. not a blocker for this PR.

Reply:
The locking was introduced by https://github.com/kiegroup/optaplanner/commit/48c9e9b5b504cdd6f91d416d3ba05292a7884788 as a fix to a rather rare race condition bug. I don't see any additional test coverage (if it's even possible and practical to develop any), though.